Fix finished_mutations_to_keep=0 for MergeTree (as docs says 0 is to keep everything)

azat opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix finished_mutations_to_keep=0 for MergeTree (as docs says 0 is to keep everything)

azat committed:
Fix finished_mutations_to_keep=0 for MergeTree (as docs says 0 is to keep everything)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-ch-test-poll3 commented:
<!-- automatic status comment for PR #60031 from azat/ClickHouse:merge_tree_mutations_cleanup-fix -->
*This is an automated comment for commit e60cf0b7666443745f05978e02b51c76a75683ab with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60031/e60cf0b7666443745f05978e02b51c76a75683ab/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60031/e60cf0b7666443745f05978e02b51c76a75683ab/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

antaljanosbenjamin commented on src/Storages/StorageMergeTree.cpp:
```diff
@@ -1463,8 +1463,11 @@ UInt64 StorageMergeTree::getCurrentMutationVersion(
 size_t StorageMergeTree::clearOldMutations(bool truncate)
 {
-    size_t finished_mutations_to_keep = truncate ? 0 : getSettings()->finished_mutations_to_keep;
+    size_t finished_mutations_to_keep = getSettings()->finished_mutations_to_keep;
+    if (!finished_mutations_to_keep)
+        return 0;
 
+    finished_mutations_to_keep = truncate ? 0 : finished_mutations_to_keep;
```
As I see we call this function with `truncate = true` when
1. the table is truncated
2. a part or partition is dropped

In case 1 I think it make sense to really truncate the old mutations, however it is not clear from the docs what is the expected behavior.

In case 2 it is questionable. To be honest I don't see a good reason why to clean the old mutations when a part or partition is droopped.

Based on these paprtial information, I have to ask: what do you think about enforcing the truncate flag regardless the value of `finished_mutations_to_keep`, but only call this function with `truncate = true` when we truncate a table?

azat replied:
Indeed, make total sense. Rebased.

_P.S. case 1 I broke unintentionally_

azat committed:
Do not clean mutations for MergeTree on DROP PART (to match ReplicatedMergeTree)

Even if this should be done, only related mutations should be cleaned,
not all.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

antaljanosbenjamin commented on src/Storages/StorageMergeTree.cpp:
```diff
@@ -1969,8 +1972,6 @@ void StorageMergeTree::dropPartition(const ASTPtr & partition, bool detach, Cont
         }
     }
 
-    /// Old parts are needed to be destroyed before clearing them from filesystem.
```
Shouldn't the same be done for `StorageMergeTree::dropPart` too? If not, then I don't see why.

azat replied:
Sure, missed that, thanks. Rebased

azat committed:
Fix expectations for test_all_projection_files_are_dropped_when_part_is_dropped

Since after DROP PART mutations are not cleaned anymore.

Here an example:

                objects_at_the_end = list_objects(cluster)
    >           assert objects_at_the_end == objects_empty_table
    E           AssertionError: assert ['data/evt/iczupcswcatzvjikqwmovahturdht', 'data/oxq/dmkgqbdzeeiwmukogoxawfoxxnrqs'] == ['data/oxq/dmkgqbdzeeiwmukogoxawfoxxnrqs']
    E             At index 0 diff: 'data/evt/iczupcswcatzvjikqwmovahturdht' != 'data/oxq/dmkgqbdzeeiwmukogoxawfoxxnrqs'
    E             Left contains one more item: 'data/oxq/dmkgqbdzeeiwmukogoxawfoxxnrqs'
    E             Full diff:
    E               [
    E             +  'data/evt/iczupcswcatzvjikqwmovahturdht',
    E                'data/oxq/dmkgqbdzeeiwmukogoxawfoxxnrqs',
    E               ]

    test_replicated_zero_copy_projection_mutation/test.py:155: AssertionError

And decoded paths:

    node1 :) select local_path from system.blob_storage_log where remote_path = 'data/evt/iczupcswcatzvjikqwmovahturdht'

    SELECT local_path
    FROM system.blob_storage_log
    WHERE remote_path = 'data/evt/iczupcswcatzvjikqwmovahturdht'

    Query id: 9ee5a9c0-c3b7-46ad-82bd-64c8bcbda78d

    ┌─local_path────────────────────────────────────────────────────────┐
    │ store/bce/bcea71c9-35cd-4368-9504-c563253b1964/tmp_mutation_1.txt │
    │ store/bce/bcea71c9-35cd-4368-9504-c563253b1964/mutation_1.txt     │
    └───────────────────────────────────────────────────────────────────┘

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>Bugfix validation — New test(s) failed to reproduce a bug

- https://github.com/ClickHouse/ClickHouse/pull/60375

>Integration tests (release) [3/4] — fail: 1, passed: 641

- `test_replicated_zero_copy_projection_mutation/test.py::test_all_projection_files_are_dropped_when_part_is_dropped` - since now mutations are note removed on DROP PART. I've updated the test

antaljanosbenjamin commented:
Apart from the stress test all the failures are flaky tests:
- `02782_uniq_exact_parallel_merging_bug` [CI DB](https://play.clickhouse.com/play?user=play#U0VMRUNUIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMsIGNoZWNrX3N0YXR1cywgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICBBTkQgcHVsbF9yZXF1ZXN0X251bWJlciA9IDAKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnU0tJUFBFRCcKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnT0snCiAgICBBTkQgY2hlY2tfc3RhdHVzICE9ICdzdWNjZXNzJwogICAgQU5EIHRlc3RfbmFtZSA9ICcwMjc4Ml91bmlxX2V4YWN0X3BhcmFsbGVsX21lcmdpbmdfYnVnJwpPUkRFUiBCWSBjaGVja19zdGFydF90aW1lIGRlc2MsIGNoZWNrX25hbWUsIHRlc3RfbmFtZQ==)
- `01459_manual_write_to_replicas` [CI DB](https://play.clickhouse.com/play?user=play#U0VMRUNUIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMsIGNoZWNrX3N0YXR1cywgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICBBTkQgcHVsbF9yZXF1ZXN0X251bWJlciA9IDAKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnU0tJUFBFRCcKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnT0snCiAgICBBTkQgY2hlY2tfc3RhdHVzICE9ICdzdWNjZXNzJwogICAgQU5EIHRlc3RfbmFtZSA9ICcwMTQ1OV9tYW51YWxfd3JpdGVfdG9fcmVwbGljYXMnCk9SREVSIEJZIGNoZWNrX3N0YXJ0X3RpbWUgZGVzYywgY2hlY2tfbmFtZSwgdGVzdF9uYW1l)
- `01171_mv_select_insert_isolation_long` [CI DB](https://play.clickhouse.com/play?user=play#U0VMRUNUIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMsIGNoZWNrX3N0YXR1cywgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICBBTkQgcHVsbF9yZXF1ZXN0X251bWJlciA9IDAKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnU0tJUFBFRCcKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnT0snCiAgICBBTkQgY2hlY2tfc3RhdHVzICE9ICdzdWNjZXNzJwogICAgQU5EIHRlc3RfbmFtZSA9ICcwMTE3MV9tdl9zZWxlY3RfaW5zZXJ0X2lzb2xhdGlvbl9sb25nJwpPUkRFUiBCWSBjaGVja19zdGFydF90aW1lIGRlc2MsIGNoZWNrX25hbWUsIHRlc3RfbmFtZQ==), failed twice

antaljanosbenjamin commented on tests/queries/0_stateless/02994_merge_tree_mutations_cleanup.sql.j2:
```diff
@@ -0,0 +1,14 @@
+create table data_mt (key Int) engine=MergeTree() order by tuple() settings finished_mutations_to_keep=0, merge_tree_clear_old_parts_interval_seconds=1;
+
+{% for table in ['data_rmt', 'data_mt'] %}
+alter table {{table}} delete where 1;
```
In our private CI this failed:
```
2024-03-01 07:40:58 @@ -1,4 +1,4 @@
2024-03-01 07:40:58 -mutations after ALTER for data_rmt	1
2024-03-01 07:40:58 +mutations after ALTER for data_rmt	0
2024-03-01 07:40:58  mutations after cleanup for data_rmt	1
2024-03-01 07:40:58  mutations after ALTER for data_mt	1
2024-03-01 07:40:58  mutations after cleanup for data_mt	1
```
So I think we should wait for the mutation to be executed:
```suggestion
alter table {{table}} delete where 1 settings mutations_sync = 1;
```

azat committed:
tests: wait for mutations in 02994_merge_tree_mutations_cleanup.sql.j2

Otherwise it fails in private fork CI

Co-authored-by: János Benjamin Antal <antaljanosbenjamin@users.noreply.github.com>

antaljanosbenjamin commented:
This PR looks okay, I will merge it soon. Sorry for not merging yet, but our CI started to show strange errors and I was worried to merge things without figuring those things out first.

antaljanosbenjamin commented:
Unit tests failed with:
```
[ RUN      ] ConnectionPoolTest.NoReceiveCall
/build/base/poco/Foundation/src/ThreadPool.cpp:188:14: runtime error: member call on address 0x555599f3d050 which does not point to an object of type 'Poco::Runnable'
0x555599f3d050: note: object has invalid vptr
 00 00 00 00  50 dc c7 8c 55 55 00 00  01 00 00 00 00 00 00 00  d0 9a f9 99 55 55 00 00  00 00 00 00
```

I doubt this is connected to this PR. I will open an issue about this.

`02784_parallel_replicas_automatic_decision_join` Flaky https://github.com/ClickHouse/ClickHouse/issues/60625.

clickgapai commented:
Hi @azat @antaljanosbenjamin — while reviewing this PR I found the following:

- **Coverage gap:** `TRUNCATE` clearing mutations and `DROP PARTITION` preserving mutations are untested after the `clearOldMutations` refactor — https://github.com/ClickHouse/ClickHouse/pull/103859

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.
